File: 241-Different Ways to Add Parentheses.py
class Solution(object):
    def diffWaysToCompute(self, expression):
        """
        :type expression: str
        :rtype: List[int]
        """
        memo = dict()
        
        def cal(start, end):
            if (start, end) in memo:
                return memo[(start, end)]
            if expression[start: end+1].isdigit():
                return [int(expression[start:end+1])]
            res = []
            for i in range(start, end+1):
                if expression[i] in ("+","-","*"):
                    left = cal(start, i-1)
                    right = cal(i+1, end)
                    for l in left:
                        for r in right:
                            if expression[i] == "+":
                                res.append(int(l)+int(r))
                            elif expression[i] == "-":
                                res.append(int(l)-int(r))
                            else:
                                res.append(int(l)*int(r))
            
            memo[(start, end)] = res
            return res
        
        return cal(0, len(expression)-1)

File: JavaScript/241-Different Ways to Add Parentheses.js
/**
 * @param {string} expression
 * @return {number[]}
 */
var diffWaysToCompute = function(expression) {
    let dp = new Array(21).fill().map(()=>new Array());

    let cal = (start, end) =>{
        if (dp[start][end] != undefined){
            return dp[start][end];
        }
        
        // isNaN(num) returns true if the variable does NOT contain a valid number
        
        if (!isNaN(expression.substring(start, end+1))){
            return [parseInt(expression.substring(start, end+1))];
        }
        
        let res = [];
        for (let i=start; i<=end; i++){
            const current = expression[i];
            if (isNaN(parseInt(current,10))){
                let left = cal(start, i-1), right = cal(i+1, end);
                for (let l of left){
                    for (let r of right){
                        if (current === "+"){
                            res.push(parseInt(l) + parseInt(r));
                        }else if (current === "-"){
                            res.push(parseInt(l) - parseInt(r));
                        }else{
                            res.push(parseInt(l) * parseInt(r));
                        }
                    }
                }
            }
        }
        dp[start][end] = res;
        return res;
    }
    return cal(0, expression.length-1);
};